File: cpp/tests/test_recall_message.cpp
// 定时消息撤回（RECALL_MESSAGE 370）离线单测 —— 不需要集群。
//
// 对齐基准是 Java 5.5.1 的 RecallMessageHandle / RecallMessageRequestHeader /
// DefaultMQProducerImpl#recallMessage(:1570-1601)。盯的是三类错得很安静的东西：
//   * 句柄编解码：分隔符是**空格**、版本段是 "v1"、编码是 base64url **带 '=' 填充**；
//     Java 解码器不吃无填充串，本端口两种都吃（见头文件差异说明），所以两种都要测。
//   * 报文键名：RecallMessageRequestHeader 在 Java 里继承 RpcRequestHeader，
//     brokerName 的**反射名是 bname**，写成 brokerName 会被 broker 静默丢掉。
//   * 客户端校验顺序：状态 → checkTopic → 禁 retry/DLQ → 解句柄 → 定位 broker，
//     前三步都不该打网络（一个手滑的句柄不该耗掉一次 RPC 超时）。
//
// 语义（撤回后消息到底还不投递）只有真机能证，见 examples/recall_live.cpp。
#include <chrono>
#include <cstdio>
#include <functional>
#include <string>

#include "rocketmq/client/exception.h"
#include "rocketmq/client/producer.h"
#include "rocketmq/common/recall_message_handle.h"
#include "rocketmq/remoting/protocol/codes.h"
#include "rocketmq/remoting/protocol/headers.h"

using namespace rocketmq;

namespace {

int fails = 0;
int checks = 0;

void expect(bool ok, const std::string& name, const std::string& detail = "") {
    ++checks;
    if (!ok) {
        ++fails;
        std::printf("FAIL %s %s\n", name.c_str(), detail.c_str());
    }
}

const char* kTopic = "TopicRecallUnit";
const char* kBroker = "broker-a";
const char* kUniqKey = "0123456789ABCDEF0123456789abcdef";
// Java 的 RecallMessageHandle 在所有解码失败分支上给的都是这一句文案。
const char* kInvalid = "recall handle is invalid";

std::string catchMessage(const std::function<void()>& fn) {
    try {
        fn();
    } catch (const MQClientException& e) {
        return e.what();
    } catch (const std::exception& e) {
        return std::string("other:") + e.what();
    }
    return "<no exception>";
}

// ---------------------------------------------------------------- 句柄编解码
void testCodec() {
    // 向量是用 Java 的 Base64.getUrlEncoder() 实算出来的，改动即失配。
    const std::string handle = buildRecallHandle(kTopic, kBroker, "1700000000000", kUniqKey);
    expect(handle == "djEgVG9waWNSZWNhbGxVbml0IGJyb2tlci1hIDE3MDAwMDAwMDAwMDAg"
                     "MDEyMzQ1Njc4OUFCQ0RFRjAxMjM0NTY3ODlhYmNkZWY=",
           "buildHandle 与 Java 向量一致", handle);
    expect(handle.size() % 4 == 0, "buildHandle 保留 Java 的 '=' 填充");

    const HandleV1 parsed = decodeRecallHandle(handle);
    expect(parsed.topic == kTopic && parsed.brokerName == kBroker &&
               parsed.timestampStr == "1700000000000" && parsed.messageId == kUniqKey,
           "句柄往返一致", parsed.topic + "/" + parsed.brokerName);
    // Java 的 getUrlDecoder 会拒无填充串；本端口必须两种都吃（另外三个移植用无填充编码器）。
    std::string unpadded = handle;
    while (!unpadded.empty() && unpadded.back() == '=') unpadded.pop_back();
    expect(decodeRecallHandle(unpadded) == parsed, "无填充句柄同样能解");

    // Java split(" ") 后只取 items[1..4]，多余分段直接忽略。
    // 下面是 base64url("v1 TopicA broker-a 1700000000000 abc junk")。
    const HandleV1 truncated =
        decodeRecallHandle("djEgVG9waWNBIGJyb2tlci1hIDE3MDAwMDAwMDAwMDAgYWJjIGp1bms=");
    expect(truncated.topic == "TopicA" && truncated.messageId == "abc",
           "多余分段被忽略（与 Java 一致）", truncated.messageId);

    expect(catchMessage([] { decodeRecallHandle(""); }) == kInvalid, "空句柄被拒");
    expect(catchMessage([] { decodeRecallHandle("not-a-handle"); }) == kInvalid,
           "非 v1 五段的句柄被拒");
    expect(catchMessage([] { decodeRecallHandle("!!!!"); }) == kInvalid,
           "非法 base64 字符被拒");
    // base64url("v2 TopicA b 1 id") / ("v1 TopicA b 1")
    expect(catchMessage([] { decodeRecallHandle("djIgVG9waWNBIGIgMSBpZA=="); }) == kInvalid,
           "版本号不对被拒");
    expect(catchMessage([] { decodeRecallHandle("djEgVG9waWNBIGIgMQ=="); }) == kInvalid,
           "段数不足被拒");
    // utf-8 编码的 0xFF 0xFE 开头 —— 解出来根本不是文本
    expect(catchMessage([] { decodeRecallHandle("__4gYmFkIHV0Zjg="); }) == kInvalid,
           "解出来不是合法 utf-8 时拒绝");
}

// ---------------------------------------------------------------- 报文头
void testHeaders() {
    RecallMessageRequestHeader header;
    header.producerGroup = std::string("PG");
    header.topic = std::string(kTopic);
    header.recallHandle = std::string("djEg");
    header.bname = std::string(kBroker);
    const PropertyMap ext = header.toExtFields();
    expect(ext.size() == 4 && ext.count("producerGroup") == 1 && ext.count("topic") == 1 &&
               ext.count("recallHandle") == 1 && ext.count("bname") == 1,
           "撤回请求头用 Java 的键名（brokerName 反射名是 bname）");
    expect(ext.at("bname") == kBroker, "bname 落盘", ext.count("bname") ? ext.at("bname") : "");

    RecallMessageRequestHeader decoded;
    decoded.fromExtFields(ext);
    expect(decoded.bname == kBroker && decoded.recallHandle == "djEg" && decoded.topic == kTopic &&
               decoded.producerGroup == "PG",
           "撤回请求头往返一致");

    RecallMessageResponseHeader resp;
    PropertyMap respExt;
    respExt["msgId"] = std::string(kUniqKey);
    resp.fromExtFields(respExt);
    expect(resp.msgId == kUniqKey, "撤回响应头读 msgId");
    expect(resp.toExtFields().at("msgId") == kUniqKey, "撤回响应头写 msgId");

    SendMessageResponseHeader send;
    PropertyMap sendExt;
    sendExt["msgId"] = std::string("0000000000000000");
    sendExt["recallHandle"] = std::string("djEg");
    send.fromExtFields(sendExt);
    expect(send.recallHandle == "djEg", "SEND 响应头解析 recallHandle");
    expect(send.toExtFields().at("recallHandle") == "djEg", "recallHandle 能编回线上格式");
    SendMessageResponseHeader plain;
    plain.fromExtFields(PropertyMap{{"msgId", std::string("x")}});
    expect(!plain.recallHandle.has_value(), "普通消息没有 recallHandle");
    expect(plain.toExtFields().count("recallHandle") == 0, "缺失时不会被编进报文");

    expect(RequestCode::RECALL_MESSAGE == 370, "请求码 370");
}

// ---------------------------------------------------------------- 生产者本地校验
void testProducerValidation() {
    DefaultMQProducer notStarted("PG_not_started");
    expect(catchMessage([&] { (void)notStarted.recallMessage(kTopic, "djEg"); }) ==
               "producer not started, call start() first",
           "未 start 就撤回会被拒");

    // 起了实例但地址服务器不可达：本地校验必须在打网络之前跑完。
    DefaultMQProducer p("PG_recall_unit");
    p.setNamesrvAddr("127.0.0.1:1");
    p.start();

    expect(catchMessage([&] { (void)p.recallMessage("%RETRY%PG_recall_unit", "djEg"); }) ==
               "topic is not supported",
           "%RETRY% topic 在本地就被拒（Java 文案）");
    expect(catchMessage([&] { (void)p.recallMessage("%DLQ%PG_recall_unit", "djEg"); }) ==
               "topic is not supported",
           "%DLQ% topic 在本地就被拒（Java 文案）");
    expect(catchMessage([&] { (void)p.recallMessage("bad topic!", "djEg"); }) !=
               "<no exception>",
           "非法 topic 名被拒");

    const auto began = std::chrono::steady_clock::now();
    const std::string corrupt = catchMessage([&] { (void)p.recallMessage(kTopic, "not-a-handle"); });
    const auto cost = std::chrono::duration_cast<std::chrono::milliseconds>(
                          std::chrono::steady_clock::now() - began)
                          .count();
    expect(corrupt == kInvalid, "非法句柄本地即拒", corrupt);
    expect(cost < 200, "非法句柄是秒回的（没打网络）", std::to_string(cost) + "ms");

    // 句柄合法但拿不到路由：Java 的 tryToFindTopicPublishInfo **异常照抛**
    // （DefaultMQProducerImpl:1586），所以这里 surfacing 的是路由错误而不是
    // "The broker service address not found"（后者要有路由、只是没有可用 broker 才会走到）。
    const std::string handle = buildRecallHandle(kTopic, kBroker, "1700000000000", kUniqKey);
    const std::string noRoute = catchMessage([&] { (void)p.recallMessage(kTopic, handle); });
    expect(noRoute == "Can not find Message Queue for topic: " + std::string(kTopic),
           "路由缺失时预热带异常照抛（Java 语义）", noRoute);

    p.shutdown();
}

}  // namespace

int main() {
    testCodec();
    testHeaders();
    testProducerValidation();
    std::printf("%s recall_message: %d checks, %d failures\n", fails == 0 ? "OK" : "FAILED",
                checks, fails);
    return fails == 0 ? 0 : 1;
}

File: python/tests/test_recall_message.py
# -*- coding: utf-8 -*-
"""定时消息撤回（RECALL_MESSAGE 370）离线单测，不需要集群。

对齐基准是 Java 5.5.1 的 ``RecallMessageHandle`` / ``RecallMessageRequestHeader`` /
``DefaultMQProducerImpl#recallMessage``。盯的是三类错得很安静的东西：
  * 句柄编解码：分隔符是**空格**、版本段是 ``v1``、编码是 base64url **带 ``=`` 填充**；
    Java 解码器不吃无填充串，这里两种都吃（见模块 docstring），所以两种都要测。
  * 报文键名：``RecallMessageRequestHeader`` 在 Java 里继承 ``RpcRequestHeader``，
    brokerName 的**反射名是 bname**，写成 ``brokerName`` 会被 broker 静默丢掉。
  * 客户端校验顺序：状态 → checkTopic → 禁 retry/DLQ → 解句柄 → 定位 broker，
    前三步都不该打网络。
"""
from __future__ import annotations

import time
from typing import List, Optional

import pytest

from rocketmq.client.exception import MQBrokerException, MQClientException
from rocketmq.client.mq_client import MQClientInstance
from rocketmq.client.mq_client import TopicPublishInfo
from rocketmq.client.producer import DefaultMQProducer
from rocketmq.common import recall_message_handle
from rocketmq.remoting.protocol.codes import RequestCode, ResponseCode
from rocketmq.remoting.protocol.headers import (RecallMessageRequestHeader,
                                                RecallMessageResponseHeader,
                                                SendMessageResponseHeader)
from rocketmq.remoting.protocol.remoting_command import RemotingCommand

TOPIC = "TopicRecallUnit"
BROKER = "broker-a"
UNIQ_KEY = "0123456789ABCDEF0123456789abcdef"
ADDR = "127.0.0.1:10911"


# ---------------- 句柄编解码 ----------------
def test_build_handle_matches_java_vector():
    """向量是拿 Java 的 Base64.getUrlEncoder() 实算出来的，改动即失配。"""
    handle = recall_message_handle.build_handle(TOPIC, BROKER, "1700000000000", UNIQ_KEY)
    assert handle == ("djEgVG9waWNSZWNhbGxVbml0IGJyb2tlci1hIDE3MDAwMDAwMDAwMDAg"
                      "MDEyMzQ1Njc4OUFCQ0RFRjAxMjM0NTY3ODlhYmNkZWY=")  # noqa: E501


def test_build_handle_keeps_java_padding():
    """Java 用带填充的 getUrlEncoder；无填充会让 broker 侧某些实现解不出来。"""
    handle = recall_message_handle.build_handle("a", "b", "1", "c")
    assert len(handle) % 4 == 0


def test_round_trip_and_padded_and_unpadded_both_decode():
    handle = recall_message_handle.build_handle(TOPIC, BROKER, "1700000000000", UNIQ_KEY)
    assert recall_message_handle.decode_handle(handle) == recall_message_handle.HandleV1(
        TOPIC, BROKER, "1700000000000", UNIQ_KEY)
    assert recall_message_handle.decode_handle(handle.rstrip("=")) == \
        recall_message_handle.decode_handle(handle)


def test_extra_segments_are_ignored_like_java_split():
    raw = " ".join(["v1", TOPIC, BROKER, "1700000000000", UNIQ_KEY, "whatever"])
    import base64
    handle = base64.urlsafe_b64encode(raw.encode("utf-8")).decode("ascii")
    got = recall_message_handle.decode_handle(handle)
    assert (got.topic, got.broker_name, got.timestamp_str, got.message_id) == \
        (TOPIC, BROKER, "1700000000000", UNIQ_KEY)


def _b64(text: str) -> str:
    import base64
    return base64.urlsafe_b64encode(text.encode("utf-8")).decode("ascii")


@pytest.mark.parametrize("bad", [
    "",                                             # Java: StringUtils.isEmpty
    "not-a-handle",                                 # 解出来不是 v1 开头的 5 段
    "!!!!",                                         # 非法 base64 字符 -> DecoderException
    _b64("v2 TopicA broker-a 1 " + UNIQ_KEY),       # 版本不对
    _b64("v1 TopicA broker-a 1"),                   # 段数不足
    "__4gYmFkIHV0Zjg=",                             # 解出来不是合法 utf-8
])
def test_bad_handles_fail_with_the_java_message(bad):
    with pytest.raises(MQClientException) as exc:
        recall_message_handle.decode_handle(bad)
    assert str(exc.value) == "recall handle is invalid"


# ---------------- 报文头 ----------------
def test_recall_request_headers_use_java_keys():
    header = RecallMessageRequestHeader()
    header.producer_group = "PG"
    header.topic = TOPIC
    header.recall_handle = "h"
    header.bname = BROKER
    assert header.to_ext_fields() == {
        "producerGroup": "PG", "topic": TOPIC, "recallHandle": "h", "bname": BROKER,
    }
    other = RecallMessageRequestHeader()
    other.from_ext_fields(header.to_ext_fields())
    assert other.bname == BROKER


def test_recall_response_header_and_send_response_recall_handle():
    resp = RecallMessageResponseHeader()
    resp.from_ext_fields({"msgId": UNIQ_KEY})
    assert resp.msg_id == UNIQ_KEY
    assert resp.to_ext_fields() == {"msgId": UNIQ_KEY}

    sent = SendMessageResponseHeader()
    sent.from_ext_fields({"msgId": "0" * 32, "recallHandle": "djEg"})
    assert sent.recall_handle == "djEg"
    assert sent.to_ext_fields()["recallHandle"] == "djEg"
    # 普通消息没有这个字段：既不编出来，也不误读成空串。
    assert "recallHandle" not in SendMessageResponseHeader().to_ext_fields()
    plain = SendMessageResponseHeader()
    plain.from_ext_fields({"msgId": "0" * 32})
    assert plain.recall_handle is None


# ---------------- MQClientInstance.recall_message ----------------
class _StubClient(MQClientInstance):
    """跳过 __init__（不建 socket），只留 _invoke_sync 缝，验证 370 的响应处理。"""

    def __init__(self, response: RemotingCommand):
        self.calls: List[tuple] = []
        self.response = response

    def _invoke_sync(self, addr, request, timeout_millis=None):
        self.calls.append((addr, request, timeout_millis))
        return self.response


def _recall_response(code: int, ext: Optional[dict] = None, remark: str = "") -> RemotingCommand:
    cmd = RemotingCommand.create_response_command(code, remark, None)
    cmd.ext_fields = ext or {}
    return cmd


def _request_header() -> RecallMessageRequestHeader:
    header = RecallMessageRequestHeader()
    header.producer_group = "PG"
    header.topic = TOPIC
    header.recall_handle = "djEg"
    header.bname = BROKER
    return header


def test_recall_message_rpc_returns_msg_id_and_uses_code_370():
    client = _StubClient(_recall_response(ResponseCode.SUCCESS, {"msgId": UNIQ_KEY}))
    assert client.recall_message(ADDR, _request_header(), 3000) == UNIQ_KEY
    addr, request, timeout = client.calls[0]
    assert (addr, timeout) == (ADDR, 3000)
    assert request.code == RequestCode.RECALL_MESSAGE
    request.make_custom_header_to_net()          # 真正落盘到 ext_fields 的那一步
    assert request.ext_fields == {"producerGroup": "PG", "topic": TOPIC,
                                  "recallHandle": "djEg", "bname": BROKER}


def test_recall_message_surfaces_broker_code_and_remark():
    client = _StubClient(_recall_response(ResponseCode.ILLEGAL_OPERATION,
                                          remark="recall failed, timestamp invalid"))
    with pytest.raises(MQBrokerException) as exc:
        client.recall_message(ADDR, _request_header(), 3000)
    assert exc.value.response_code == ResponseCode.ILLEGAL_OPERATION
    assert "timestamp invalid" in exc.value.error_message


def test_recall_message_rejects_success_without_msg_id():
    client = _StubClient(_recall_response(ResponseCode.SUCCESS, {}))
    with pytest.raises(MQBrokerException):
        client.recall_message(ADDR, _request_header(), 3000)


# ---------------- 生产者校验顺序 ----------------
def _producer() -> DefaultMQProducer:
    p = DefaultMQProducer("PG")
    p.name_server_addrs = [ADDR]
    return p


def test_recall_before_start_fails_locally():
    with pytest.raises(MQClientException):
        _producer().recall_message(TOPIC, "djEg")


class _NoIoClient:
    """任何网络入口都被记下来：本地校验必须在打网络**之前**跑完。"""

    def __init__(self):
        self.touched: List[str] = []
        # False = 模拟「路由拿不到」；True = 预热成功，让流程继续往下走。
        self.no_publish_error = False
        # MQClientInstance 上的真字段：拿不到路由时 validateNameServerSetting 靠它
        # 分辨"没配 name server"(10004) 和"这个 topic 没路由"(10005)，这里给个地址。
        self.name_server_addrs: List[str] = [ADDR]

    def register_topic_in_use(self, topic):
        self.touched.append("register")

    def get_topic_publish_info(self, topic, is_default=False):
        self.touched.append("publish")
        if not self.no_publish_error:
            raise MQClientException("No route info of this topic: %s" % topic)
        return TopicPublishInfo()

    def get_topic_route_data(self, topic):
        self.touched.append("route")
        return None

    def broker_addr_of(self, broker_name):
        self.touched.append("addr")
        return None

    def recall_message(self, addr, header, timeout):
        self.touched.append("rpc")
        return UNIQ_KEY


def _started(client) -> DefaultMQProducer:
    p = _producer()
    p._started = True
    p._mq_client = client
    return p


def test_retry_and_dlq_topics_are_refused_before_any_io():
    for topic in ("%RETRY%PG", "%DLQ%PG"):
        client = _NoIoClient()
        with pytest.raises(MQClientException) as exc:
            _started(client).recall_message(topic, "djEg")
        assert str(exc.value) == "topic is not supported"
        assert client.touched == []


def test_illegal_topic_names_are_refused_before_any_io():
    client = _NoIoClient()
    with pytest.raises(MQClientException):
        _started(client).recall_message("bad topic!", "djEg")
    assert client.touched == []


def test_corrupt_handle_fails_before_any_io():
    client = _NoIoClient()
    began = time.monotonic()
    with pytest.raises(MQClientException) as exc:
        _started(client).recall_message(TOPIC, "not-a-handle")
    assert str(exc.value) == "recall handle is invalid"
    assert client.touched == []
    assert (time.monotonic() - began) < 0.2


def test_missing_broker_address_uses_java_message():
    client = _NoIoClient()
    client.no_publish_error = True          # 路由预热成功，才轮到定位 broker 地址
    handle = recall_message_handle.build_handle(TOPIC, BROKER, "1700000000000", UNIQ_KEY)
    with pytest.raises(MQClientException) as exc:
        _started(client).recall_message(TOPIC, handle)
    assert str(exc.value) == "The broker service address not found"
    # 路由预热先于定位地址，且一次网络都没打。
    assert "publish" in client.touched
    assert "rpc" not in client.touched


def test_publish_info_failure_propagates_like_java():
    """DefaultMQProducerImpl:1586 的 tryToFindTopicPublishInfo 没有 try/catch：
    路由拿不到就直接失败，不会继续往后拼请求。"""
    client = _NoIoClient()
    handle = recall_message_handle.build_handle(TOPIC, BROKER, "1700000000000", UNIQ_KEY)
    with pytest.raises(MQClientException) as exc:
        _started(client).recall_message(TOPIC, handle)
    assert "No route info" in str(exc.value)
    assert "addr" not in client.touched and "rpc" not in client.touched


def test_recall_goes_out_to_the_broker_named_in_the_handle():
    client = _NoIoClient()
    client.no_publish_error = True
    client.broker_addr_of = lambda broker_name: ADDR      # type: ignore[assignment]
    handle = recall_message_handle.build_handle(TOPIC, BROKER, "1700000000000", UNIQ_KEY)
    assert _started(client).recall_message(TOPIC, handle) == UNIQ_KEY
    assert client.touched[-1] == "rpc"
